### PR TITLE
 T5 Gradient Checkpointing

### LINKED ISSUE #6564: T5 Gradient Checkpointing
# 🚀 Feature request

Currently, only Bert supports gradient checkpointing which allow the model to be fine-tuned on GPUs with small memory.
It will be great to make T5 also support gradient checkpointing.

Code:
https://github.com/huggingface/transformers/blob/0735def8e1200ed45a2c33a075bc1595b12ef56a/src/transformers/modeling_bert.py#L461

## Motivation

T5 has very big models with 3B and 11B parameters which make it impossible to be fine-tuned on most GPUs. Gradient checkpointing will allow these huge models to be fine-tuned on GPUs. This will lead to much better results on downstream tasks using on house GPUs without the need to fine-tuned it on TPUs.

## Your contribution

If I am not mistaken all what need to be change is the following block:
https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_t5.py#L752

```
for i, (layer_module, past_key_value_state) in enumerate(zip(self.block, past_key_value_states)):
            if output_hidden_states:
                all_hidden_states = all_hidden_states + (hidden_states,)

            if getattr(self.config, "gradient_checkpointing", False):

                def create_custom_forward(module):
                    def custom_forward(*inputs):
                        return module(*inputs, output_attentions)

                    return custom_forward

                 layer_outputs = torch.utils.checkpoint.checkpoint(
                    create_custom_forward(layer_module),
                    hidden_states,
                    extended_attention_mask,
                    position_bias,
                    encoder_hidden_states,
                    encoder_extended_attention_mask,
                    encoder_decoder_position_bias,
                    head_mask[i],
                    past_key_value_state,
                    use_cache,
                    output_attentions,
                )

            else:
                layer_outputs = layer_module(
                    hidden_states,
                    attention_mask=extended_attention_mask,
                    position_bias=position_bias,
                    encoder_hidden_states=encoder_hidden_states,
                    encoder_attention_mask=encoder_extended_attention_mask,
                    encoder_decoder_position_bias=encoder_decoder_position_bias,
                    head_mask=head_mask[i],
                    past_key_value_state=past_key_value_state,
                    use_cache=use_cache,
                    output_attentions=output_attentions,
                )
                # layer_outputs is a tuple with:
                # hidden-states, key-value-states, (self-attention weights), (self-attention position bias), (cross-attention weights), (cross-attention position bias)
            hidden_states, present_key_value_state = layer_outputs[:2]

            if i == 0:
                # We share the position biases between the layers - the first layer store them
                # layer_outputs = hidden-states, key-value-states (self-attention weights), (self-attention position bias), (cross-attention weights), (cross-attention position bias)
                position_bias = layer_outputs[3 if output_attentions else 2]
                if self.is_decoder and encoder_hidden_states is not None:
                    encoder_decoder_position_bias = layer_outputs[5 if output_attentions else 3]
            # append next layer key value states
            if use_cache:
                present_key_value_states = present_key_value_states + (present_key_value_state,)

            if output_attentions:
                all_attentions = all_attentions + (layer_outputs[2],)  # We keep only self-attention weights for now
```

@patrickvonplaten  thanks in advance for looking into it.

### PR DESCRIPTION
**Pull request by ceshine**
# What does this PR do?

Partially fixes #6564. This is inspired by @xFinal's workaround. However, instead of modifying PyTorch's implementation of gradient checkpointing, I only modify `T5Block` here and replace the None value with a dummy tensor that requires a gradient.

~Gradient checkpointing is only enabled for the encoder. Some of the outputs of the decoder don't require gradients and will cause problems. PyTorch 1.8.0 has fixed this (see note 1 below).~

Additional notes:

1. `require_grad = True` for the dummy tensor is no longer required since PyTorch 1.8.0 ([From this PR](https://github.com/pytorch/pytorch/pull/45934)).
2. None as a return value is allowed since [this PyTorch PR](https://github.com/pytorch/pytorch/pull/52422). It has not been released yet (the latest release is 1.8.1 at the time of writing). We won't even need the dummy tensor after that.
3. I tested this code locally with PyTorch 1.7.1.
4. I did not write any additional test because it seems that [test_training_gradient_checkpointing](https://github.com/huggingface/transformers/blob/81009b7a5c5cb183a9275c15bf347bdc988b02c4/tests/test_modeling_common.py#L242) in `ModelTesterMixin` already covers it.

~EDIT: I forgot this PR only covers the encoder, not the decoder. The description has been updated to reflect this fact.~
EDIT: This PR covers the decoder now.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation). (Updated docstring for T5Config.)
- [ ] Did you write any new necessary tests? (Already covered by existing tests.)

## Who can review?

@patrickvonplaten (T5)
@sgugger (Documentation)

### CODE REVIEW

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -689,6 +690,12 @@ def forward(
 
         outputs = (hidden_states,)
 
+        if present_key_value_state is None and self.training:
```
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_t5.py:693]: I think it would be nicer if we can slightly change the code to **not** even return `present_key_value_state` if `use_cache=False`. (You could check how it's done for GPT2, e.g. :-))

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -945,23 +952,61 @@ def forward(
...
-                encoder_hidden_states=encoder_hidden_states,
-                encoder_attention_mask=encoder_extended_attention_mask,
-                encoder_decoder_position_bias=encoder_decoder_position_bias,
-                layer_head_mask=layer_head_mask,
-                encoder_layer_head_mask=encoder_layer_head_mask,
-                past_key_value=past_key_value,
-                use_cache=use_cache,
-                output_attentions=output_attentions,
-            )
+            if (
```
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_t5.py:955]: nice this looks good!

File: src/transformers/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -945,23 +952,61 @@ def forward(
...
+                getattr(self.config, "gradient_checkpointing", False)
+                and self.training
+                and (not self.config.is_decoder)
```
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_t5.py:958]: I guess it could/should also work for the decoder no?

Comment (ceshine) [reply src/transformers/models/t5/modeling_t5.py:958]: For the encoder, we only need to deal with the None return value (`present_key_value_state`). Things are a bit more complicated with the decoder. Sometimes some of the returned tensors have `requires_grad=False`. I don't think there is a perfect way to deal with that before PyTorch 1.8. If we enable gradients for those tensors, there will be some unnecessary computation. I also don't know how to do that yet. IMO, we should use another PR to deal with the decoder.

File: src/transformers/models/t5/modeling_t5.py (hunk 1 above)
Comment (ceshine) [reply src/transformers/models/t5/modeling_t5.py]: Yes, it's indeed a cleaner approach. I've updated the code.

File: src/transformers/models/t5/modeling_t5.py (hunk 3 above)
Comment (ceshine) [reply src/transformers/models/t5/modeling_t5.py:958]: I've tracked down the tensor that is causing the problem. It's the position bias returned by `T5Attention` with `has_relative_attention_bias=False`. The gradient checkpointing would work if we force the position bias to require gradients when training ([as in this commit](https://github.com/veritable-tech/transformers/commit/acaeee6b2e675045fb28ce2176444c1d63e908bd)). I'm not sure about the ramification of this change. If it seems okay to you, I can incorporate this commit into this PR.

Comment (ceshine) [reply src/transformers/models/t5/modeling_t5.py:958]: We can also only apply this rule only when gradient checkpointing is on ([as in this commit](https://github.com/veritable-tech/transformers/commit/658bdd0bd1215353a8770f558bda2ea69a0ad0c7)).

File: src/transformers/models/t5/modeling_t5.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/t5/modeling_t5.py]: Thanks for updating!

File: src/transformers/models/t5/modeling_t5.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/models/t5/modeling_t5.py:958]: Thanks for investigating! 

Adding this change:

```python
                if self.training and self.gradient_checkpointing:
                    position_bias.requires_grad = True
```

sounds great to me!

Comment (ceshine) [reply src/transformers/models/t5/modeling_t5.py]: Awesome! I've incorporated the changes in this PR.

**Review by patrickvonplaten — APPROVED** [review]
Great job @ceshine! Pinging @patil-suraj for final review

**Review by patil-suraj — APPROVED** [review]
Amazing and really clean!

Thanks a lot for working on this!
